File: PAGES/scraper.php
<?php

include('../simple_html_dom.php');
include('../DB/ManagerFarmaco.php');
include('../DB/Farmaco.php');
ini_set('max_execution_time', 300);

$pagina = 1;
while($amafarma = file_get_html('https://www.amafarma.com/farmaci-da-banco.html?p=' . $pagina)) {
    $listaProdotti = $amafarma->find('li[class="item product product-item"]');
    foreach($listaProdotti as $farmaco) {
        $minsan = $farmaco->find('[data-product-sku]', 0)->attr['data-product-sku'];
        $prezzoNuovo = $farmaco->find('[span[class="price"]]', 0)->plaintext;
        
        // ----------------------------- CONTROLLA DELLA CONVENIENZA -----------------------------
        $farmacoConfronto = ManagerFarmaco::readById($minsan);
        if(is_bool($farmacoConfronto) == 1 || $farmacoConfronto->getPrezzo() > $prezzoNuovo) {
            $linkSito = $farmaco->find('div[class="product-item-info"] a', 0)->href;
            $dettagliSito = file_get_html($linkSito);

            $nomeProdotto = $farmaco->find('div[class="product details product-item-details"] strong[class="product name product-item-name"] a', 0)->plaintext;
            
            $img = $farmaco->find('span[class="product-image-wrapper"] img', 1)->src;

            $percorsoDescrizione = 'div[class="product attribute description"] div[class="value"]';
            if($dettagliSito->find($percorsoDescrizione, 0)) {
                $descrizione = $dettagliSito->find($percorsoDescrizione, 0)->plaintext;
            }
            $categoria = $dettagliSito->find('span[class="categoria_riferimento"] span a', 0)->plaintext;

            if($farmaco->find('span[class="price"]', 1)) {
                $prezzoVecchio = $farmaco->find('span[class="price"]', 1)->plaintext;
            }

            $prezzoNuovo = str_replace(",", "", $prezzoNuovo);
            $prezzoVecchio = str_replace(",", "", $prezzoVecchio);

            $prezzoNuovo = str_replace("€", "", $prezzoNuovo);
            $prezzoVecchio = str_replace("€", "", $prezzoVecchio);

            /*
                PROBLEMA SPAZIO DOPO I PREZZI
            */
            $farmacoNuovo = new Farmaco($minsan, $nomeProdotto, 1000, 2000, $descrizione, $img, $categoria);
            ManagerFarmaco::create($farmacoNuovo);
        }
        // ----------------------------- CONTROLLA DELLA CONVENIENZA -----------------------------
    }

    $pagina++;
}
?>

File: scraper.py
import requests
from bs4 import BeautifulSoup
import pandas as pd
headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.82 Safari/537.36'}

pagina = 1
listaFarmaci = []
arrayLink = []
contatoreFarmaci = 1

# lettura numero elementi
k = requests.get('https://www.amafarma.com/farmaci-da-banco.html').text
soup=BeautifulSoup(k,'html.parser')
farmaciTotali = int(soup.find("p", class_="toolbar-amount").findChildren()[2].text)

# scorrimento delle pagine
while contatoreFarmaci < farmaciTotali:
    k = requests.get('https://www.amafarma.com/farmaci-da-banco.html?p=' + str(pagina)).text
    soup=BeautifulSoup(k,'html.parser')
    listaFarmaci = soup.find_all("li", class_="item product product-item")

    for farmaco in listaFarmaci:
        link = farmaco.find("a", class_="product photo product-item-photo").get('href')

        prezzi = farmaco.find_all("span", class_="price")
        prezzoVecchio = prezzi[0].text        
        if len(prezzi) > 1:
            prezzoNuovo = prezzi[1].text
        nomeProdotto = farmaco.find("strong", class_="product name product-item-name").findChildren("a")[0].text.strip()

        img = farmaco.find("span", class_="product-image-wrapper").findChildren("img")[0].get("src")

        k = requests.get(link).text
        dettagliSito=BeautifulSoup(k,'html.parser')

        minsan = dettagliSito.find("span", class_="sku").findChildren()[1].text
        dettagliGenerali = dettagliSito.find("div", class_="product attribute description")
        if dettagliGenerali != None:
            descrizione = dettagliGenerali.findChildren("div", class_="value")[0].text

        categoria = dettagliSito.find("span", class_="categoria_riferimento").findChildren("a")[0].text
        contatoreFarmaci = contatoreFarmaci + 1

    pagina = pagina + 1